fio/lib/src/utils.ts

```diff
@@ -0,0 +1,281 @@
+ */
+interface ChainData {
+  chain_id: string;
+  expiration: string;
+  ref_block_num: number;
+  ref_block_prefix: number;
+}
+
+/**
+ * Represents an ABI provider.
+ */
+interface AbiProvider {
+  getRawAbi: () => Promise<{ accountName: string; abi: Uint8Array }>;
+}
+
+/**
+ * Represents the ABI for the FIO token.
+ */
+interface FioTokenAbi {
+  account_name: string;
+  abi: string;
+  abi_hash: string;
+  code_hash: string;
+}
+
+/**
+ * Represents the response of pushing a transaction.
+ */
+interface PushTransactionResponse {
+  transaction_id: string;
+  processed: {
+    receipt: {
+      status: string;
+      cpu_usage_us: number;
+      net_usage_words: number;
+    };
+  };
+}
+
+/**
+ * Creates a raw transaction.
+ * @param publicKey The public key.
+ * @param actionData The action data.
+ * @param account The account.
+ * @param action The action.
+ * @returns The raw transaction, serialized action data, and chain ID.
+ */
+export const createRawTx = async ({
+  publicKey,
+  actionData,
+  account,
+  action,
+  fioBaseUrl,
+}: CreateRawTxParams): Promise<{
+  rawTx: any;
+  serializedActionData: string;
+  chainId: string;
+}> => {
+  const chainData: ChainData = await getChainData(fioBaseUrl);
+  Transactions.FioProvider = {
+    accountHash: Fio.accountHash,
+  };
+
+  const { fioTokenAbi } = await getAbiProvider(account);
+  Transactions.abiMap.set(fioTokenAbi.account_name, fioTokenAbi);
+
+  const serializedActionData = await createSerializeActionData({
+    account,
+    actionName: action,
+    actionData,
+    chainId: chainData.chain_id,
+  });
+
+  const transaction = new Transactions();
+  const rawTx = await transaction.createRawTransaction({
+    action: action,
+    account: account,
+    data: actionData,
+    chainData: chainData,
+    publicKey: publicKey,
+  });
+
+  return {
+    rawTx,
+    serializedActionData,
+    chainId: chainData.chain_id,
+  };
+};
+
+/**
+ * Serializes action data.
+ * @param account The account.
+ * @param actionName The action name.
+ * @param actionData The action data.
+ * @param chainId The chain ID.
+ * @returns The serialized action data.
+ */
+const createSerializeActionData = async ({
+  account,
+  actionName,
+  actionData,
+  chainId,
+}: CreateSerializeActionDataParams): Promise<string> => {
+  const { abiProvider } = await getAbiProvider(account);
+  // @ts-ignore
+  const fioApi = new Api({
+    chainId: chainId,
+    abiProvider: abiProvider,
+  });
+  const contract = await fioApi.getContract(account);
+  // @ts-ignore
+  const serializedActionData = serializeActionData(
+    contract,
+    account,
+    actionName,
+    actionData
+  );
+  return serializedActionData;
+};
+
+/**
+ * Gets the ABI provider and FIO token ABI.
+ * @param account The account.
+ * @returns The FIO token ABI and ABI provider.
+ */
+const getAbiProvider = async (
+  account: string
+): Promise<{ fioTokenAbi: FioTokenAbi; abiProvider: AbiProvider }> => {
```

Comment:
I think you should split this in 2 functions.

getFioAddressAccountAbi
getAbiProvider

Reply:
or have the map as a file const and use that to get the ABIProvider.

Reply:
Done